domain/src/test/java/org/oppia/domain/oppialogger/exceptions/OppiaUncaughtExceptionHandlerTest.kt

```diff
@@ -0,0 +1,203 @@
+
+  @Inject
+  lateinit var networkConnectionUtil: NetworkConnectionUtil
+
+  @Inject
+  lateinit var fakeDefaultExceptionHandler: FakeDefaultExceptionHandler
+
+  @Inject
+  lateinit var exceptionsController: ExceptionsController
+
+  @Inject
+  lateinit var fakeExceptionLogger: FakeExceptionLogger
+
+  @InternalCoroutinesApi
+  @Inject
+  lateinit var testCoroutineDispatchers: TestCoroutineDispatchers
+
+  @Mock
+  lateinit var mockOppiaExceptionLogsObserver: Observer<AsyncResult<OppiaExceptionLogs>>
+
+  @Captor
+  lateinit var oppiaExceptionLogsResultCaptor: ArgumentCaptor<AsyncResult<OppiaExceptionLogs>>
+
+  @Before
+  fun setUp() {
+    networkConnectionUtil = NetworkConnectionUtil(ApplicationProvider.getApplicationContext())
+    setUpTestApplicationComponent()
+    Thread.setDefaultUncaughtExceptionHandler(fakeDefaultExceptionHandler)
+  }
+
+  @ExperimentalCoroutinesApi
+  @InternalCoroutinesApi
+  @Test
+  fun testHandler_throwException_withNoNetwork_verifyLogInCache() {
+    networkConnectionUtil.setCurrentConnectionStatus(NetworkConnectionUtil.ConnectionStatus.NONE)
+    val exceptionThrown = Exception("TEST")
+    oppiaUncaughtExceptionHandler.uncaughtException(Thread.currentThread(), exceptionThrown)
+
+    val cachedExceptions = exceptionsController.getExceptionLogStore()
+    dataProviders.convertToLiveData(cachedExceptions).observeForever(mockOppiaExceptionLogsObserver)
+    testCoroutineDispatchers.advanceUntilIdle()
+
+    verify(mockOppiaExceptionLogsObserver, atLeastOnce())
+      .onChanged(oppiaExceptionLogsResultCaptor.capture())
+
+    val exceptionLog = oppiaExceptionLogsResultCaptor.value.getOrThrow().getExceptionLog(0)
+    val exception = exceptionLog.toException()
+
+    assertThat(exception.message).matches("java.lang.Exception: TEST")
+  }
+
+  @Test
+  fun testHandler_throwException_withNetwork_verifyLogToRemoteService() {
+    val exceptionThrown = Exception("TEST")
+    oppiaUncaughtExceptionHandler.uncaughtException(Thread.currentThread(), exceptionThrown)
+
+    val exceptionCaught = fakeExceptionLogger.getMostRecentException()
+    assertThat(exceptionCaught).hasMessageThat().matches("java.lang.Exception: TEST")
+    assertThat(exceptionCaught.cause).isEqualTo(Exception(exceptionThrown).cause)
+  }
+
+  private fun setUpTestApplicationComponent() {
+    DaggerOppiaUncaughtExceptionHandlerTest_TestApplicationComponent.builder()
+      .setApplication(ApplicationProvider.getApplicationContext())
+      .build()
+      .inject(this)
+  }
+
+  @Test
+  fun testHandler_throwException_verifyLogToDefaultHandler(){
+    val exceptionThrown = Exception("TEST")
+    oppiaUncaughtExceptionHandler.uncaughtException(Thread.currentThread(), exceptionThrown)
+
+    val exceptionCaught = fakeDefaultExceptionHandler.getMostRecentException()
+    assertThat(exceptionCaught).hasMessageThat().matches("java.lang.Exception: TEST")
+    assertThat(exceptionCaught.cause).isEqualTo(Exception(exceptionThrown).cause)
```

Comment:
Why do we need to do Exception(exceptionThrown) here? Can we not just do exceptionThrown.cause?

Reply:
We need to do Exception(exceptionThrown) because we're doing Exception(throwable) in our handler while logging to our controller. Our controller needs an exception, so we transform the throwable into an exception.